src/testdrive/asciidoc.py

```diff
@@ -1,104 +1,459 @@
-        '|===',
-        '',
-        '==== Results',
-        '',
-        '[%header,cols=5*.^a]',
-        '|===',
-        '|id|timestamp|duration (s)|result|reason',
-        '',
-    ]
-    for case in cases:
-        if case['result'] is True:
-            result = 'success'
-        elif case['result'] is False:
-            result = 'failure'
-        else:
-            result = case['result']
-        lines.append(row(
-            case['id'],
-            case.get('timestamp', '_not known_'),
-            case.get('time', '_not known_'),
-            result,
-            case['reason'] or '',
-        ))
-    lines.append('|===')
-    return '\n'.join(lines)
+        Any image files will be copied to a unique filename in `objdir`/images.
+        """
+        for (title, path) in self._images:
+            filename = f'{uuid4()}{os.path.splitext(path)[1]}'
+            copyfile(path, os.path.join(objdir, 'images', filename))
+            yield ''
+            yield f'.{title or os.path.basename(path)}'
+            yield f'image::{filename}[]'
+        for (title, dct) in self._tables:
+            yield ''
+            yield f'.{title}'
+            yield '[cols="1,4"]'
+            yield '|==='
+            yield ''
+            yield from (row(f'*{k}*', dct[k]) for k in sorted(dct))
+            yield ''
+            yield '|==='
+    @staticmethod
+    def _image_item(item):
+        """Return (title, path) for the image specified by `item`.
+
+        Return None if `item` is not a mapping specifying a file 'path' and,
+        optionally, a 'title', or is not a string (specifying a file path).
+        """
+        try:
+            return (item.get('title'), item['path'])
+        except (AttributeError, TypeError):
+            if not isinstance(item, str):
+                return None
+            return (None, item)
+    @classmethod
+    def from_output(cls, output):
+        """Return an instance of `cls` if `output` is JSON-encoded test detail.
+
+        Return None if `output` is not a JSON-encoded object or does not contain
+        test detail (as understood by this class).
+        """
+        required = {'result', 'reason'}
+        try:
+            obj = json.loads(output)
+            if frozenset(obj.keys()).intersection(required) != required:
+                return None
+        except (TypeError, json.JSONDecodeError, AttributeError):
+            return None
+        images = []
+        for item in obj.get('plot', ()):
+            try:
+                (title, path) = cls._image_item(item)
+            except ValueError:
+                return None
+            images.append((title, path))
+        tables = []
+        dct = obj.get('analysis')
+        if dct is not None:
+            if not isinstance(dct, dict):
+                return None
+            analysis = {}
+            for (key, val) in dct.items():
+                if isinstance(val, dict):
+                    tables.append((key, val))
```

Comment:
Let's take the PMC example, how this is going to be represented in the report?
```json
{
  "analysis": {
    "duration": 1999.0688657,
    "clock_class_count": {
      "FREERUN": {
        "count": 0,
        "transitions": {
          "FREERUN": 0,
          "LOCKED": 0,
          "HOLDOVER_IN_SPEC": 0,
          "HOLDOVER_OUT_SPEC1": 0,
          "HOLDOVER_OUT_SPEC2": 0,
          "HOLDOVER_OUT_SPEC3": 0
        }
      }
    }
  }
}
```

As I see it, it should be a table for every state. Will the `clock_class_count` be represented also?

Reply:
A table won't be built for the example you posted. As an unhandled case, this will result in the whole of stdout (your full JSON) being rendered as a literal block instead.

Reply:
The question we need to answer first is how to represent your data in asciidoc. If we have N states, do we want N tables (2 columns = 1 for state names + 1 for counts, N rows) or 1 table (an NxN matrix of counts)?

Once we have answered that, I think we can decide what layout in JSON.

Reply:
I think N tables could be good enough. A table for every state with transition count

Reply:
In my JSON example I would expect it to look like something like this
```adoc
.analysis
[%autowidth]
[cols="1,1"]
|===
| duration | 1999.0688657
|===

.FREERUN
[%autowidth]
[cols="1, 2a"]
|===
| count | 0
| transitions
| 
[cols="2,1"]
!===
! State ! Number Of Transitions

! FREERUN
! 0
! LOCKED
! 0
! HOLDOVER_IN_SPEC
! 0
! HOLDOVER_OUT_SPEC1
! 0
! HOLDOVER_OUT_SPEC2
! 0
! HOLDOVER_OUT_SPEC3
! 0
!===
|===
```

Reply:
Carrying this proposal forward outside of this PR.